goldmane/pkg/internal/utils/files.go

```diff
@@ -15,73 +15,80 @@ package utils
 // to the given channel when any of the files change.
-func WatchFilesFn(updChan chan struct{}, files ...string) (func(context.Context), error) {
-	fileWatcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return nil, fmt.Errorf("error creating file watcher: %s", err)
-	}
+//
+// Note that we don't use fsnotify or a similar library here - the way that file projection in Kubernetes works relies on
+// a number of temporary files being created, copied, and removed, which can pose problems for some file watching libraries.
+// We generally don't need immediate updates on file changes, so we poll the files at a regular interval instead.
+func WatchFilesFn(updChan chan struct{}, interval time.Duration, files ...string) (func(context.Context), error) {
+	cached := map[string]string{}
 
-	watchedDirs := map[string]struct{}{}
 	for _, file := range files {
-		if file == "" {
-			continue
-		}
-
-		// There is a bug in fsnotify where some events are not triggered on files. So watch the directory -
-		// this ensures we catch all events. Longer term, moving off of fsnotify would be a good idea.
-		dir := filepath.Dir(file)
-		if _, ok := watchedDirs[dir]; ok {
-			// Already watching this directory.
-			continue
-		}
-
-		if err := fileWatcher.Add(dir); err != nil {
-			logrus.WithError(err).Warn("Error watching directory for changes")
-			continue
-		}
-		logrus.WithField("dir", dir).Debug("Watching directory for changes")
-		watchedDirs[dir] = struct{}{}
+		logrus.WithField("file", file).Debug("Watching file")
 	}
 
 	return func(ctx context.Context) {
-		for dir := range watchedDirs {
-			logrus.WithField("dir", dir).Info("Starting watch on directory")
-		}
-
-		// If we exit this function, make sure to close the file watcher and update channel.
-		defer fileWatcher.Close()
+		// If we exit this function, make sure to close the update channel.
 		defer close(updChan)
 		defer logrus.Warn("File watcher closed")
+
 		for {
-			select {
-			case <-ctx.Done():
+			if ctx.Err() != nil {
 				return
-			case event, ok := <-fileWatcher.Events:
-				if !ok {
-					logrus.Warn("File watcher events channel closed")
-					return
+			}
+
+			for _, file := range files {
+				logCtx := logrus.WithField("file", file)
+
+				// Read the file.
+				data, err := os.ReadFile(file)
+				if err != nil {
+					logrus.WithError(err).Errorf("Failed to read file %s", file)
+					if errors.Is(err, os.ErrNotExist) {
+						delete(cached, file)
+					}
+					if !chanutil.WriteNonBlocking(updChan, struct{}{}) {
+						logCtx.Debug("file notification channel is full, dropping update")
+					}
+					continue
 				}
-				if !chanutil.WriteNonBlocking(updChan, struct{}{}) {
-					logrus.WithField("event", event).Debug("file notification channel is full, dropping update")
+
+				// Hash the file.
+				hash := fmt.Sprintf("%x", sha256.Sum256(data))
+
+				// Compare the hash to the cached hash.
+				if cur, ok := cached[file]; !ok {
```

Comment:
Do we really want to skip a notification if a file was added? I think it could be that it's not configured to start, maybe a blank config map was mounted, then it's added later when it's configured?

We could start mounting empty secrets / config maps to avoid restarts for things like this.

Reply:
Yeah, I had this thought as well and we probably do want to generate an event. But this made sure that the existing UTs passed :sweat_smile: